src/write/encoder.rs

```diff
@@ -89,38 +86,58 @@ impl<'a, W: Write> fmt::Debug for EncoderWriter<'a, W> {
     }
 }
 
-impl<'a, W: Write> EncoderWriter<'a, W> {
+impl<W: Write> EncoderWriter<W> {
     /// Create a new encoder that will write to the provided delegate writer `w`.
-    pub fn new(w: &'a mut W, config: Config) -> EncoderWriter<'a, W> {
+    pub fn new(w: W, config: Config) -> EncoderWriter<W> {
         EncoderWriter {
             config,
-            w,
+            delegate: Some(w),
             extra_input: [0u8; MIN_ENCODE_CHUNK_SIZE],
             extra_input_occupied_len: 0,
             output: [0u8; BUF_SIZE],
             output_occupied_len: 0,
-            finished: false,
             panicked: false,
         }
     }
 
     /// Encode all remaining buffered data and write it, including any trailing incomplete input
     /// triples and associated padding.
     ///
-    /// Once this succeeds, no further writes can be performed, as that would produce invalid
-    /// base64.
+    /// Once this succeeds, no further writes or calls to this method are allowed.
     ///
-    /// This may write to the delegate writer multiple times if the delegate writer does not accept all input provided
-    /// to its `write` each invocation.
+    /// This may write to the delegate writer multiple times if the delegate writer does not accept
+    /// all input provided to its `write` each invocation.
+    ///
+    /// If you don't care about error handling, it is not necessary to call this function, as the
+    /// equivalent finalization is done by the Drop impl.
+    ///
+    /// Returns the writer that this was constructed around.
     ///
     /// # Errors
     ///
-    /// The first error that is not of [`ErrorKind::Interrupted`] will be returned.
-    pub fn finish(&mut self) -> Result<()> {
-        if self.finished {
-            return Ok(());
+    /// The first error that is not of `ErrorKind::Interrupted` will be returned.
+    pub fn finish(&mut self) -> Result<W> {
```

Comment:
I think the more idiomatic action would be to call this `into_inner()` and make it take `self`. This also removes the need for `Option<W>` in the body. However that would be a less subtle API break and would be quite likely to cause issues.

Also note that this is an API break because of the return type. People who were using it in a position where the value was being "used" (probably after a `?`) will get an error now because `W != ()`.

Reply:
Another option here is to keep `finish` with the previous signature and add an `into_inner()` method with the new signature. It wouldn't allow you to get rid of the `Option` (and associated complexity) but it might make the change fully backwards compatible. You can then could mark `finish` as deprecated and one day remove it (and the `Option`).

Reply:
Ah you are right! I had actually done this first and then undid it (see comment below) but finish() is _not_ retryable because it has to have an internal loop to keep writing, which means that interrupted errors are already ignored. So, problem solved there.

Re naming: to me I wouldn't expect into_inner() to perform I/O, but maybe that precedent is well laid elsewhere?

Reply:
There is precedent for IO in std: https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.into_inner

Reply:
I'm okay with the pretty minor compatibility load of adding a `let _ = `. I would bet that all but the most careful people aren't calling finish() to begin with, and are just letting `Drop` take care of it. Hopefully the careful people won't mind...

Reply:
Doh -- turns out I still need the `Option`, as I cannot move the `W` out of `EncoderWriter` since it implements `Drop`.

Isn't it breaking API, regardless of return type, to make `finish` take `self`?

Reply:
It's up to you. Personally I would prefer deprecating this to make it take self by value but you are probably right that few if any users will break and that the fix is very easy. At least for my use I am relying on `Drop` so I have no horse in the race.

Reply:
> Doh -- turns out I still need the `Option`, as I cannot move the `W` out of `EncoderWriter` since it implements `Drop`.

Ah right. I forget about this rust "bug".

> Isn't it breaking API, regardless of return type, to make `finish` take `self`?

It definitely is. You would want to make a different method.

Reply:
I think I was wrong with my first comment here: finish() is retryable because my policy of retrying on interrupt is just that: _my_ policy (based on write_all's logic of nebulous validity). Another user might decide that other `ErrorKind`s are retryable in their use case, so it would be inappropriate to consume `self` if `finish()` returned `Err` that they might want to retry on.

Reply:
I don't think there is any sort of general guarantee that `write_all` (or `write`) don't do partial writes (or full writes but return an error anyways). I guess since the user knows what `Write` implementation they are using they can be smart about the retries if they make assumptions about how this type works.

So you are right, in the general case retries are not safe, but depending on what the user chooses to assume about this crate and the `Write` impl they can meaningfully retry.

Reply:
With `into_inner` out of reach since it can't consume `self`, got any ideas for an expressive name for `finish but also return the delegate writer`? If there's something that's not too odious, you've convinced me that that's worth the trouble to stay compatible.

Reply:
I can't think of a good name. I think if it can't be made to consume `self` than it is less attractive to make a new method.